[Enhancement] Improve feedback and grading generation

schultek opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I updated multiple integration tests (Spring) related to the features (with a high test coverage)

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Preparing the grading and feedback system of programming exercises to be extended with the new static code analysis configuration of #2086.

### Description
<!-- Describe your changes in detail -->

- Cleanup and refactor feedback generation in bamboo and jenkins service.
    - Moved generation of test-case-feedback from BambooService to FeedbackService (now in-line with sca feedback generation)
    - Remove retrieveLatestResult method used for the 'trigger-failed-build' endpoint
        - the retrieved result had different information than the result sent to Artemis via the updated bamboo notification plugin
        - therefore the score generation would be false
        - the endpoint now **always** triggers a new build
- Add ProgrammingExerciseGradingService and ProgrammingExerciseGradingResource to handle grading specific api requests and score generation
    - *Most of the code is moved and adapted from the TestCaseService / -Resource classes.*
    - Moved **processNewProgrammingExerciseResult** from ResultService to GradingService
    - Moved **updateResult** and **updateAllResults** from TestCaseService to GradingService
    - Moved **reset** and **re-evaluate** endpoints from TestCaseResource to GradingResource
- **We do not yet include static code analysis in the score generation**.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

Check that the grading system for programming exercises still works as before.
Since this affects CI-specific code, please test this for both setups - Bamboo and Jenkins (TS1 / TS2).

1. Log in to Artemis
2. Create a Programming Exericse
3. Adjust the TestCase Configuration via the Grading page.
4. Participate in the exercise
5. Check that the score is calculated correctly and all test case feedback is displayed correctly.

Also test:
- Re-evaluating the results via the grading page.
- Triggering all builds via the grading page.
- Resetting the grading config via the grading page.
- Triggering a failed build as a student (Stop the build on bamboo/jenkins and refresh the page to see the red button).
- Tests after due date
- Sequential tests
- Test cases with bonus Points
- Participation after due date should still be possible with an unrated result
- Manual results (inline feedback feature)

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->

- FeedbackService.java: 95%
- ProgrammingExerciseGradingService.java: 94%
- ProgrammingExerciseTestCaseService.java: 99%
- ProgrammingExerciseGradingResource.java: 97%
- ProgrammingExerciseTestCaseResource.java: 94%

schultek commented:
> ![Codacy](https://camo.githubusercontent.com/b72992711a855bd9a03bfbdfc5ef7fcf528482f5/68747470733a2f2f6170702e636f646163792e636f6d2f6173736574732f696d616765732f66617669636f6e2e706e67) Here is an overview of what got changed by this pull request:
> 
> ```diff
> Issues
> ======
> + Solved 13
> - Added 2
>            
> 
> Coverage increased per file
> ===========================
> + src/main/java/de/tum/in/www1/artemis/exception/BambooException.java  25
> + src/main/java/de/tum/in/www1/artemis/service/connectors/BitbucketBambooUpdateService.java  9
> + src/main/java/de/tum/in/www1/artemis/service/connectors/ContinuousIntegrationService.java  17
> + src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java  14
> + src/main/java/de/tum/in/www1/artemis/web/websocket/programmingSubmission/BuildTriggerWebsocketError.java  66
> + src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java  5
> + src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java  4
>          
> 
> Coverage decreased per file
> ===========================
> - src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/dto/BambooTestResultDTO.java  -100
> - src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/dto/BambooArtifactsDTO.java  -100
> - src/main/java/de/tum/in/www1/artemis/service/ResultService.java  -3
> - src/main/java/de/tum/in/www1/artemis/service/FileService.java  -1
> - src/main/java/de/tum/in/www1/artemis/domain/enumeration/StaticCodeAnalysisTool.java  -21
> - src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java  -2
> - src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/dto/QueriedBambooBuildResultDTO.java  -100
> - src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/dto/BambooChangesDTO.java  -43
> - src/main/java/de/tum/in/www1/artemis/service/FeedbackService.java  -6
>          
> 
> Complexity increasing per file
> ==============================
> - src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseGradingResource.java  2
> - src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java  9
> - src/main/java/de/tum/in/www1/artemis/exception/ContinousIntegrationException.java  1
>          
> 
> Complexity decreasing per file
> ==============================
> + src/main/java/de/tum/in/www1/artemis/service/ResultService.java  -2
> + src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java  -6
> + src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java  -1
>          
> ```
> 
> See the complete overview on [Codacy](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=6189331&bid=20195594)

I was surprised by the amount of decreased coverage, so here is why:
Most of the decreased coverage of the bamboo files relate to the changed handling of trigger-failed-builds. Since this does not directly fetch the latest build from bamboo anymore the tests have changed accordingly. Previously this request was mocked using all the DTO's from above. So this is why there aren't tested anymore.

There is still a use of the 'fetchLatestBuildResult' method which is in the 'getParticipationBuildArtifact' endpoint. But this is just tested on a shallow level without using the DTO's. Improving this might be useful, but is not in scope of this PR.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/FeedbackService.java:
```diff
@@ -94,6 +97,59 @@ public FeedbackService(FeedbackRepository feedbackRepository) {
+     * @param programmingLanguage the programming language of the exercise.
+     * @return Feedback object for the test job
+     */
+    public Feedback createFeedbackFromTestJob(BambooBuildResultNotificationDTO.BambooTestJobDTO testJob, boolean successful, final ProgrammingLanguage programmingLanguage) {
```
Important: FeedbackService (and all other services) must be independent of Bamboo specifics. It should only depend on the more general `ContinuousIntegrationService`. This is an important design principle in Artemis to avoid issues with other VC / CI systems (Gitlab, Jenkins) and future implementations of those.

I think you have two options:
1) Move this method into Bamboo service
2) Find an abstraction for BambooTestJobDTO that can be used for Jenkins, Bamboo and potentially all other CI Services

The 2nd option is used for the StaticCodeAnalysisReportDTO which is basically independent from the concrete CI system. However, this does not mean that you also need to choose it here

schultek replied:
Ok good point. I think I'm just using the name and errors property of the BambooTestJobDTO. I could also just give these directly as parameters to the method.

schultek replied:
I updated the signature of the method which is now independent of bamboo-specific models.
Therefore it is now also used in the JenkinsService.

krusche commented:
I think the testing instructions are not detailed enough.

As you removed the method `retrieveLatestBuildResult` and changed the logic here, the case `trigger-failed-build` once (student) and `trigger-failed-build` all (instructor) need to be tested thoroughly.

By the way: just removing functionality like this is considered as bad practice. You should actively communicate about this beforehand! Several people put thoughts into this before and there is a reason that this functionality is available!

schultek commented:
> I think the testing instructions are not detailed enough.
> 
> As you removed the method `retrieveLatestBuildResult` and changed the logic here, the case `trigger-failed-build` once (student) and `trigger-failed-build` all (instructor) need to be tested thoroughly.
> 
> By the way: just removing functionality like this is considered as bad practice. You should actively communicate about this beforehand! Several people put thoughts into this before and there is a reason that this functionality is available!

The change just affects the single endpoint for the students, not the instructor trigger-all endpoint.
Is there a way to force a failed build? Then I will add this to the testing instructions.

I understand the reason why this functionality was implemented, but as for the new system with static code analysis this would lead to a severe bug in the system, which is why i removed it. I also discussed this change with @kloessst beforehand.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/FeedbackService.java:
```diff
@@ -94,6 +96,60 @@ public FeedbackService(FeedbackRepository feedbackRepository) {
+     * @param programmingLanguage the programming language of the exercise.
+     * @return Feedback object for the test job
+     */
+    public Feedback createFeedbackFromTestJob(String testName, List<String> testErrors, boolean successful, final ProgrammingLanguage programmingLanguage) {
```
The method name is now misleading as there is no test job any more.
I suggest to use a different one

schultek replied:
createFeedbackFromTestCase ?

kloessst reviewed (dismissed):
Tested all scenarios on TS1 and TS2. Worked perfectly for me. The code looks good and the refactoring is reasonable. Well done!

kloessst reviewed (dismissed):
Re-approve

anditurdiu reviewed (changes requested):
I'm not really sure what went wrong, there are no error messages in the console. I created a programming exercise and then tried to access "Configure Grading":
![image](https://user-images.githubusercontent.com/15110960/94921557-d6d97080-04b8-11eb-9927-e2e9eaec9672.png)
It shows no data to display: 
![image](https://user-images.githubusercontent.com/15110960/94921590-e658b980-04b8-11eb-84fd-38002d3e856e.png)
I then tried it again, with another exercise, following the same procedure and for this one it does show the different tests:
![image](https://user-images.githubusercontent.com/15110960/94921655-01c3c480-04b9-11eb-957c-fcb76856af91.png)
![image](https://user-images.githubusercontent.com/15110960/94921681-0ee0b380-04b9-11eb-8480-87af6f43d193.png)

I checked bitbucket and the tests are present for both exercises.

Apart from this issue, everything worked as expected including participating, grading, and changing the grading for the second exercise + reevaluating.

EDIT: Nevermind, it was a local caching issue, after doing a hard refresh the tests were shown as expected.

anditurdiu reviewed (approved):
Everything works as expected and the code looks good too. Nice work!
Tested it again with 4 others to include all scenarios

sleiss commented:
![image](https://user-images.githubusercontent.com/5084100/94940464-d69a9e80-04d3-11eb-8509-0f3245626e15.png)
TS2 when trying to save

sleiss reviewed (approved):
Tested it on TS2, worked fine (except some unrelated TypeScript errors).

MaisiKoleni commented on src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java:
```diff
@@ -361,64 +362,24 @@ public boolean buildPlanIdIsValid(String projectKey, String buildPlanId) {
 
-            return feedback;
-        }).collect(Collectors.toList());
+        report.getResults().forEach(testsuite -> testsuite.getTestCases().forEach(testCase -> {
```
Is there a reason why we use two nested `forEach` calls here instead of the `flatMap` and `collect` that was there before?

schultek replied:
Because of the result.addFeedback() method we don't need to .stream() anymore. I did this to be consistent with the BambooService

schultek replied:
Changed it to normal for each loops.

MaisiKoleni reviewed (changes requested):
Still works as far as I can see, tested on TS 1.

Left some in-line comments for the code; apart from some small issues it looks good to me.

schultek commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java:
```diff
@@ -0,0 +1,379 @@
+                .collect(Collectors.toList());
+        feedbacksToFilterForCurrentDate.forEach(result::removeFeedback);
+        // If there are no feedbacks left after filtering those not valid for the current date, also setHasFeedback to false.
+        if (result.getFeedbacks().stream().noneMatch(feedback -> feedback.isPositive() == Boolean.FALSE
```
Changed this.

schultek commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java:
```diff
@@ -0,0 +1,379 @@
+     */
+    private Predicate<ProgrammingExerciseTestCase> isSuccessful(Result result) {
+        return testCase -> result.getFeedbacks().stream()
+                .anyMatch(feedback -> feedback.getText() != null && feedback.getText().equals(testCase.getTestName()) && feedback.isPositive() == Boolean.TRUE);
```
Changed this.

schultek commented on src/main/java/de/tum/in/www1/artemis/service/FeedbackService.java:
```diff
@@ -94,6 +96,60 @@ public FeedbackService(FeedbackRepository feedbackRepository) {
+    private String processResultErrorMessage(final ProgrammingLanguage programmingLanguage, final String message) {
+        if (programmingLanguage == ProgrammingLanguage.JAVA) {
+            // Splitting string at the first linebreak to only get the first line of the Exception
+            return message.split("\\n", 2)[0]
```
I added a TODO comment for this.

schultek commented on src/main/java/de/tum/in/www1/artemis/exception/ContinousIntegrationException.java:
```diff
@@ -0,0 +1,24 @@
+package de.tum.in.www1.artemis.exception;
+
+/**
+ * Created by muenchdo on 22/06/16.
```
I remove the comment.

anditurdiu reviewed (approved):
reapprove

fde312 commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java:
```diff
@@ -0,0 +1,379 @@
+            }
+            result = updateResult(result, programmingExercise, !isSolutionParticipation && !isTemplateParticipation);
+            result = resultRepository.save(result);
+            // workaround to prevent that result.submission suddenly turns into a proxy and cannot be used any more later after returning this method
```
I think the comment is misplaced here

fde312 commented on src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java:
```diff
@@ -705,26 +630,23 @@ private Result addFeedbackToResultNew(Result result, List<BambooBuildResultNotif
-                    String methodName = successfulTest.getName(); // in the attribute "methodName", bamboo seems to apply some unwanted logic
-
-                    createAutomaticFeedback(result, methodName, true, null);
+                    result.addFeedback(feedbackService.createFeedbackFromTestCase(successfulTest.getName(), successfulTest.getErrors(), true, programmingLanguage));
```
I would prefer to extract the service call and store it in one variable. It makes it more readable. You did the same a few lines below :)

fde312 commented:
For me everything worked fine on ts1. On ts2 it shows that my build failed even though I just added `Collections.sort(input)` see:
<img width="1670" alt="image" src="https://user-images.githubusercontent.com/41108487/94983004-5f025900-053f-11eb-9142-dd37c9018078.png">
<img width="1658" alt="image" src="https://user-images.githubusercontent.com/41108487/94983079-bef8ff80-053f-11eb-94d9-524a52a50b11.png">
<img width="1637" alt="image" src="https://user-images.githubusercontent.com/41108487/94983081-c7513a80-053f-11eb-8f17-e4858d736763.png">

Does this happen to one of the others also on ts2?
